widgetssdk/src/main/java/com/glia/widgets/view/VisitorCodeView.kt

```diff
@@ -68,60 +97,43 @@ class VisitorCodeView internal constructor(
         }
     }
 
-    override fun setController(controller: VisitorCodeContract.Controller?) {
-        this.controller = controller
-        this.controller?.setView(this)
+    override fun setController(controller: VisitorCodeContract.Controller) {
+        this.controller = controller!!
+        this.controller.setView(this)
     }
 
     private fun setDefaultTheme(typedArray: TypedArray) {
         theme = Utils.getThemeFromTypedArray(typedArray, this.context)
         applyDialogTheme(theme)
     }
 
-    fun setVisitorCodeRepository(visitorCodeRepository: VisitorCodeRepository) {
-        this.visitorCodeRepository = visitorCodeRepository
-        loadNewVisitorCode()
-    }
-
     private fun loadNewVisitorCode() {
-        notifyOfLoading()
-        if (visitorCodeRepository == null) {
-            val error: Exception = IllegalStateException("Missing visitor code repository")
-            notifyOfError("Internal error, invalid view setup", error)
-            return
-        }
-
-        visitorCodeRepository?.getVisitorCode { visitorCode, error ->
-            if (error != null) {
-                notifyOfError("Failed to load Visitor Code", error)
-                return@getVisitorCode
-            }
-            if (visitorCode == null || visitorCode.code == null) {
-                val error: Exception = IllegalStateException("Empty Visitor Code returned by the Glia Core")
-                notifyOfError("Failed to load Visitor Code", error)
-                return@getVisitorCode
-            }
-
-            notifyOfNewVisitorCode(visitorCode)
-        }
+        startLoading()
+        controller.onLoadVisitorCode()
     }
 
-    private fun notifyOfLoading() {
+    override fun startLoading() {
         runOnUi {
             showProgressBar(true)
+            showFirstAndHideSecond(successContainer, failureContainer)
         }
     }
 
-    private fun notifyOfError(publicErrorMessage: String, exception: Throwable?) {
-        Logger.e(TAG, publicErrorMessage, exception)
+    override fun onError(message: String, throwable: Throwable) {
+        Logger.e(TAG, message, throwable)
 
         runOnUi {
             showProgressBar(false)
-            charCodeView.setText("ERROR") // Will be replace and improved in the next ticket MOB-1823
+            showFirstAndHideSecond(failureContainer, successContainer)
         }
     }
 
-    private fun notifyOfNewVisitorCode(visitorCode: VisitorCode) {
+    private fun showFirstAndHideSecond(show: View, hide: View) {
```

Comment:
this function looks confusing to me, as an idea - could be better to have `showSuccess` and `showFailure` functions. e.g. 
```
fun showSuccess() {
    failureContainer.isVisible = false
    successContainer.isVisible = true
}
```

Reply:
done